From 02ca9187190f8e01deaf86e11bdc9925c2ea2a8c Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Wed, 30 Oct 2024 23:58:27 +0000
Subject: [PATCH] Fix incorrect logging in sdp_discovery

In two locations in sdp_discovery.cc, log statements use structures that
may, in exceptional cases, have been freed by preceding calls.  This can
lead to use after free and potentially to security vulnerabilities.

Use local variables instead.

Bug: 375408314
Bug: 375409435
Test: m libbluetooth
Test: researcher POC
Flag: EXEMPT trivial logic fix
Ignore-AOSP-First: security
Tag: #security
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:bb2f54f9ed938267c2830da4a9d984529274d8a8)
Merged-In: I1c8816721588acba9110257b61b4e993cf2720d6
Change-Id: I1c8816721588acba9110257b61b4e993cf2720d6
---
 system/stack/sdp/sdp_discovery.cc | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/system/stack/sdp/sdp_discovery.cc b/system/stack/sdp/sdp_discovery.cc
index 7e8bacab80..251d2fc94e 100644
--- a/system/stack/sdp/sdp_discovery.cc
+++ b/system/stack/sdp/sdp_discovery.cc
@@ -180,7 +180,7 @@ static void sdp_snd_service_search_req(tCONN_CB* p_ccb, uint8_t cont_len, uint8_
   if (stack::l2cap::get_interface().L2CA_DataWrite(p_ccb->connection_id, p_cmd) !=
       tL2CAP_DW_RESULT::SUCCESS) {
     log::warn("Unable to write L2CAP data peer:{} cid:{} len:{}", p_ccb->device_address,
-              p_ccb->connection_id, p_cmd->len);
+              p_ccb->connection_id, p - p_start);
   }
 
   /* Start inactivity timer */
@@ -708,7 +708,7 @@ static void process_service_search_attr_rsp(tCONN_CB* p_ccb, uint8_t* p_reply,
     if (stack::l2cap::get_interface().L2CA_DataWrite(p_ccb->connection_id, p_msg) !=
         tL2CAP_DW_RESULT::SUCCESS) {
       log::warn("Unable to write L2CAP data peer:{} cid:{} len:{}", p_ccb->device_address,
-                p_ccb->connection_id, p_msg->len);
+                p_ccb->connection_id, p - p_start);
     }
 
     /* Start inactivity timer */
-- 
2.34.1

